dvda-author.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.43.3.
.TH DVDA-AUTHOR "1" "August 2019" "dvda-author version dev-9999" "User Commands"
.SH NAME
dvda-author \- manual page for dvda-author version dev-9999
.SH SYNOPSIS
.B dvda-author
[\fIOPTION\fR]...
.SH DESCRIPTION
dvda\-author dev\-9999 creates high\-resolution DVD\-Audio discs
.PP
from .wav, .flac and other audio files.
.SH OPTIONS

Output options
.PP
\fB\-h\fR, \fB\-\-help\fR               Display this help.
.PP
\fB\-v\fR, \fB\-\-version\fR            Display version.
.PP
\fB\-q\fR, \fB\-\-quiet\fR              Quiet mode.
.PP
\fB\-d\fR, \fB\-\-debug\fR              Increased verbosity (debugging level)
.PP
\fB\-t\fR, \fB\-\-veryverbose\fR        Like \fB\-d\fR with enhanced verbosity for sample counts.
.TP
, \fB\-\-maxverbose\fR
Like \fB\-t\fR with maximum verbosity on audio buffers (devel info).
.TP
\fB\-\-no\-output\fR
Does not produce any file structure except for \fB\-\-fixwav\fR.
.IP
Computations will be performed.
.PP
\fB\-P\fR, \fB\-\-pause\fR              Insert a final pause before exiting.
.PP
\fB\-P0\fR, \fB\-\-pause\fR=\fI0\fR           Suppress a final pause before exiting
.IP
if specified in configuration file.
.PP
\fB\-l\fR, \fB\-\-log\fR  path          Ouput a log to filepath.
.IP
Argument must be supplied.
.PP
\fB\-L\fR, \fB\-\-logrefresh\fR path    Same as \fB\-l\fR but prior log will be erased on launching again.
.TP
\fB\-\-loghtml\fR
Transform the log into HTML format, with colorized messages, adjacent to log with extension .html added.
.PP
\fB\-k\fR, \fB\-\-text\fR               Generates text table in IFO files
.IP
Under development, implemented for 1\-group discs.
.IP
Use file information as arguments separated by commas.
.PP
\fB\-W\fR, \fB\-\-disable\-lexer\fR      Deactivates configuration file parsing.
.PP
Playback options
.PP
\fB\-a\fR, \fB\-\-autoplay\fR           Launches playback on loading disc.
.PP
Authoring options
.IP
Soundfile authoring
.PP
Supported audio types:   .wav
.IP
\&.flac and .oga (Ogg FLAC, see below)
.IP
SoX\-supported formats with \fB\-S\fR enabled
.IP
except for lossy formats.
.TP
\fB\-\-project\fR [file]
Launches a dvda\-author project.
.IP
The project file should have the same syntax as dvda\-author.conf [see file HOW_TO.conf]
.IP
By default, launches dvda\-author.dap in the current directory.
.PP
\fB\-i\fR, \fB\-\-input\fR directory    Input directory with audio files.
.IP
Each subdirectory is a group.
.PP
\fB\-o\fR, \fB\-\-output\fR directory   Output directory.
.PP
\fB\-x\fR, \fB\-\-extract\fR {disc or directory} Extract DVD\-Audio to directory \fB\-o\fR.
.IP
Groups are labelled g1, g2 in output directory.
.TP
\fB\-\-xlist\fR 1:1,...,t1n\-2:1,...,t2n\-...\-N:1,...,tNn
Optional hyphen\-separated list of groups to be extracted
.IP
may be added with \fB\-x\fR.
.IP
Tracks to be extracted in a given group are indicated after a colon.
.IP
Tracks to be extracted may be listed separated by commas after colon.
.IP
If not specified, all the group will be extracted.
.IP
Contiguous tracks may be represented by `...` between commas
.IP
e.g. 2\-3\-4:1,2,...,7,9 means: extract groups 2 and 3 entirely and tracks 2 to 7 and 9 for in group 4.
.TP
\fB\-\-aob\-extract\fR {directory or AOB[,AOB...]}
Direct AOB file audio content extraction.
.IP
Unlike \fB\-x\fR, this option does not use .IFO files. Use this option in combination with \fB\-o\fR. No wav header is generated.
.IP
Several AOB files may be listed, separated by commas.
.TP
\fB\-\-aob2wav\fR {directory or AOB[,AOB...]}
Like \fB\-\-aob\-extract\fR but a wav header is prepended to audio content.
.TP
\fB\-\-forensic\fR
Use this mode with \fB\-\-aob\-extract\fR, \fB\-\-aob2wav\fR or \fB\-x\fR if IFO files are missing or mangled, or AOB files
.IP
have been partially restored using recovery tools.
.TP
\fB\-\-strict\fR
Use this option with \fB\-\-aob\-extract\fR, \fB\-\-aob2wav\fR or \fB\-x\fR to stop extraction in case of severe header issues. May be useful in combination with \fB\-\-forensic\fR to manually repair header issues.
.TP
\fB\-\-log\-decode\fR [AOB]
Decode AOB file and log MPEG specifics. Should be used only in conjunction with \fB\-\-outfile\fR
.TP
\fB\-\-outfile\fR [file]
Path to the log generated by \fB\-\-log\-decode\fR. Caution : should be alone in its own directory.
.PP
\fB\-p\fR, \fB\-\-startsector\fR NNN    Specify the number of the first sector
.IP
of the AUDIO_PP.IFO file in the output of mkisofs.
.IP
If NNN=0, falling back on 281 (default).
.IP
Without \fB\-p\fR start sector will be computed automatically.
.PP
\fB\-g\fR                       You may specify up to 9 groups of tracks.
.IP
Minimum: one group.
Enter full path to files if input directory is not set
.IP
by [\-i].
.PP
\fB\-z\fR,                      BROKEN. Separate two consecutive titles when files have same audio
.IP
characteritics within a group.(
.PP
\fB\-Z\fR, \fB\-\-playlist\fR           You may specify up to 9 group copies.
.IP
Total number of groups and copy groups should not exceed 9.
.PP
\fB\-n\fR, \fB\-\-no\-videozone\fR       Do not generate an empty VIDEO_TS directory.
.PP
\fB\-w\fR, \fB\-\-rights\fR             Access rights to directories created (octal values)
.PP
\fB\-c\fR, \fB\-\-cga\fR                Enter channel group assignment right after group, e.g: \fB\-g\fR file1...fileN \fB\-c\fR cga1...cgaN
.IP
Channel assignment should match number of channels of each file
.IP
Combine channels using either decimal indexes in following table or hyphenated channel assignement labels
.IP
e.g. \fB\-g\fR a.wav \fB\-g\fR b.wav \fB\-c\fR Lf\-Rf\-C2\-Lfe2\-S2 \fB\-\-cga\fR 17
.IP
Channel group assignment (CGA)
.TP
Index
1    2        3         4        5       6
.TP
0
M
.TP
1
L     R
.TP
2
Lf    Rf      S2
.TP
3
Lf    Rf      Ls2      Rs2
.TP
4
Lf    Rf      Lfe2
.TP
5
Lf    Rf      Lfe2     S2
.TP
6
Lf    Rf      Lfe2     Ls2      Rs2
.TP
7
Lf    Rf      C2
.TP
8
Lf    Rf      C2       S2
.TP
9
Lf    Rf      C2       Ls2      Rs2
.TP
0xA\-10
Lf    Rf      C2       Lfe2
.TP
0xB\-11
Lf    Rf      C2       Lfe2     S2
.TP
0xC\-12
Lf    Rf      C2       Lfe2     Ls2      Rs2
.TP
0xD\-13
Lf    Rf      C        S2
.TP
0xE\-14
Lf    Rf      C        Ls2      Rs2
.TP
0xF\-15
Lf    Rf      C        Lfe2
.TP
0x10\-16 Lf
Rf      C        Lfe2     S2
.TP
0x11\-17 Lf
Rf      C        Lfe2     Ls2      Rs2
.TP
0x12\-18 Lf
Rf      Ls       Rs       Lfe2
.TP
0x13\-19 Lf
Rf      Ls       Rs       C2
.TP
0x14\-20 Lf
Rf      Ls       Rs       C2       Lfe2
.IP
Keys:
.IP
Index 2 means channel belongs to Group2
.IP
M: Mono
.IP
Lf: Left front
.IP
Rf: Right front
.IP
Ls: Left surround (behind)
.IP
Rs: Right front
.TP
C:
Center
.IP
Lfe: Low Frequency Effect (Subwoofer)
.IP
S: Surround (just one behind)
.TP
Ls: Left
surround
.IP
Rs: Right surround
.IP
Each group must have either same sample rate or be even multiples (e.g. 96kHz/48 kHz or 88.2 kHz/44.1 kHz)
.IP
Within groups, bitrate may differ but sample rate cannot.
.TP
\fB\-\-downmix\fR
Enter downmix coefficients in dB. If track has more than 2 channels, each channel (Lf, Rf, C, Ls or S, Rs, LFE)
.IP
will be mapped to left (l) and/or right (r) stereo channel
.IP
with volume reduced by x dB, x the channel downmix coefficient.
.IP
Enter positive dB values corresponding to each channel volume reduction, mapped to left or right stereo, separated by commas.
.IP
Schema is \fB\-\-downmix\fR Lf.l,Lf.r,Rf.l,Rf.r,C.l,C.r,S.l,S.r,Rs.l,Rs.r,LFE.l,LFE.r
.IP
Use 100 for 'off' value and 0 for no volume reduction.
.IP
This option can be repeated up to 16 times. It is cyclically recycled to 16 times to provide as many downmix tables.
.IP
Each track can be indexed using \fB\-\-dtable\fR to be downmixed with the corresponding table.
.IP
Example : \fB\-\-downmix\fR 6.2,100,100,7.2,0,0,13,13,16,16,0,0 \fB\-\-downmix\fR 5.2,100,100,8.2,0,0,10,10,12,12,0,0 \fB\-\-dtable\fR 2 \fB\-\-dtable\fR 1
.IP
means that track 1 will be downmixed using the second table and track 2 the first one.
.TP
\fB\-\-dtable\fR
Enter downmix table rank (1\-based) as indicated above.
.TP
\fB\-\-provider\fR
Enter provider name.
.PP
\fB\-F\fR, \fB\-\-fixwav\fR(options)    Bad wav headers will be fixed by fixwav.
.IP
Can be run alone without DVD\-AUDIO output by adding \fB\-\-nooutput\fR.
.PP
\fB\-f\fR, \fB\-\-fixwav\-virtual\fR(options)  Use .wav header repair utility
.IP
without any write operation.
.PP
\fB\-S\fR, \fB\-\-sox\fR                Use SoX to convert file format to .wav.
.IP
Without \fB\-S\fR or \fB\-\-resample\fR (see below), only flac, Ogg FLAC
.IP
and .wav files are accepted.
.TP
\fB\-\-resample\fR c b s
Use SoX to convert file format to .wav, and/or change channel to c, bitrate to b, samplerate to s.
.IP
You do not need to use \fB\-S\fR with \fB\-\-resample\fR.
.IP
Always provide the three values c b and s separated by white space in the same order as in the previous \fB\-g\fR files.
.TP
\fB\-\-padding\fR
Reverse default behaviour for transition between audio tracks with identical
.IP
characteristics (number of channels, bit depth and sample rate).
.IP
If necessary, audio will be padded with 0s instead of being joined (default).
.IP
Use \fB\-\-pad\-cont\fR for padding with last\-known byte.
.PP
\fB\-C\fR, \fB\-\-pad\-cont\fR           When padding, pad with last known byte, not 0. See \fB\-\-padding\fR above.
.IP
Deactivates \fB\-\-lossy\-rounding\fR
.PP
\fB\-L\fR, \fB\-\-lossy\-rounding\fR     Sample count rounding will be performed by cutting audio files
.IP
instead of padding (see \fB\-\-padding\fR and \fB\-\-pad\-cont\fR).
.IP
Deactivates \fB\-\-pad\-cont\fR and \fB\-\-padding\fR.
.PP
Menu authoring
.PP
\fB\-m\fR, \fB\-\-topmenu\fR(=mpgfiles) Generates top menu from comma\-separated list of mpgfiles.
.IP
Without argument, automatic menu generation is launched.
.PP
\fB\-u\fR, \fB\-\-duration\fR hh:mm:ss Duration of top menu file, if provided.
.IP
It is mandatory when \fB\-\-topmenu\fR has an argument file.
.PP
\fB\-M\fR, \fB\-\-xml\fR filepath       Generates dvdauthor xml project
.IP
to filepath.
.PP
\fB\-H\fR, \fB\-\-spuxml\fR filepath    Generates spumux xml project
.IP
to filepath.
.PP
\fB\-G\fR, \fB\-\-image\fR file         Menu Background image for customized menu authoring.
.PP
\fB\-E\fR, \fB\-\-highlight\fR file     Menu Highlight image for customized menu authoring.
.PP
\fB\-e\fR, \fB\-\-select\fR  file       Menu Select image
.IP
image that appears on pressing Enter with remote control
usually Background with a change in text color.
.PP
\fB\-N\fR, \fB\-\-blankscreen\fR file   For automatic menu authoring, you can replace black video background with this image.
.PP
\fB\-O\fR, \fB\-\-screentext\fR string  Text for top menu. Format is
.IP
"album_text=group1_text=text(track11),text(track21),...,text(trackn1):group2_text=text(track12),text(track22)...,text(trackn2):..."
.IP
with text(tracknk) the text for track n of group k and
.IP
groupk_text the text for group k.
.PP
\fB\-U\fR, \fB\-\-loop\fR               loop background video.
.PP
\fB\-K\fR, \fB\-\-highlightformat\fR    \fB\-1\fR for automatic menu authoring
.IP
with little square before titles, 0 for underlining, 1 for button highlight.
.PP
\fB\-J\fR, \fB\-\-font\fR a,b,c         Font name,font size,font width
.IP
(number of pixels for width of font size 10).
.IP
Font name should be compatible with Image magick specifications (mogrify \fB\-list\fR font).
.TP
\fB\-\-fontname\fR a
Font name.
.TP
\fB\-\-fontsize\fR b
Font size.
.TP
\fB\-\-fontwidth\fR b
Font width.
.PP
\fB\-Y\fR, \fB\-\-topmenu\-palette\fR string     Text for system palette. Format is
.IP
either "norefresh", to block the refreshing of menu images, or:
.IP
textcolor:highlight_color:select_action color
.IP
in alpha\-YCrCb 32\-bit hexa coding. Here textcolor is the non\-highlighted text for tracks,
.IP
,highlight_color is the underline or mobile motif color,
.IP
and select_action_color is album and group labels
.IP
as well as color of tracks on pressing the highlighted track.
.PP
\fB\-8\fR, \fB\-\-activemenu\-palette\fR string     Text for menu colors. Format is:
.IP
textcolor:highlight_text_color:highlight_color:select_action color
.IP
in alpha\-YCrCb 32\-bit hexa coding. Here textcolor is the text for tracks,
.IP

NEWS
Release History

   dvda-author-14.02
          Fev     20, 2014  libfixwav becomes obligatory build. Fix audio encoding regression 
	                    (starting 08.12)
   dvda-author-13.11
          Nov     11, 2013  Implement lplex authoring for hybrids.
          
   dvda-author-13.10
          Oct     01, 2013  Build system fixes. Audio menu enhancements. Updated dependencies.
	  Oct     19, 2013  Further build system fixes. Port to GNU make entirely, remove all Makefile.am files

   dvda-author-10.06
	  July    17, 2010  Other fixes and top menu soundtrack authoring with optional slideshows (build 300).
          June    14, 2010  Major fixes for bugs and regressions. Portability enhancements.

   dvda-author-09.09
	  Nov     30, 2009  User-defined and automatic submenus, continuous and hierarchical.
	  Oct     01, 2009  New build system and pruned source package. Release candidate 2.
	  Sep     01, 2009  Experimental version with top-menu and still picture authoring
                            and new audio options. Added to SF git repository.

   dvda-author-09.05
          May     27, 2009  Stabilized release added SoX interface.

   dvda-author-08.12
          Dec     25, 2008  First release with multichannel support and multiformat support.

   dvda-author-08.09-dev
          Sept    20, 2008  First release.

   dvda-author-08.08-dev
          August  20, 2008  Source code of 08-08-dev made portable to *nix platforms.
          August  10, 2008  First release.

   dvda-author-08.07
          Sept    01, 2008  Build 11 released.
          August  23, 2008  Build 8 released with cumulative fixes for portability
          August  02, 2008  Made available on http://dvd-audio.sourceforge.net
          July    30, 2008  Corrected bugs, see ChangeLog.

   dvda-author-08.06
          July    04, 2008  Made available on http://dvd-audio.sourceforge.net
          June    29, 2008  Finalized tests.
          June        2008  New developments by Fabrice Nicol.

   dvda-author-20050703
          July    03, 2005  Latest Dave Chapman source code.



src/dvda-author.c
/*

dvda-author2.c  - Author a DVD-Audio DVD

Copyright Fabrice Nicol <fabnicol@users.sourceforge.net> July 2008-2013

The latest version can be found at http://dvd-audio.sourceforge.net

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#ifndef __WIN32__
#include <unistd.h>
#endif
#include <getopt.h>
#include <sys/stat.h>
#include <sys/types.h>
#define off64_t long long
#include <dirent.h>
#include <locale.h>
#include <errno.h>
#include <inttypes.h>
#include <string.h>
#include "commonvars.h"
#include "structures.h"
#include "c_utils.h"
#include "audio2.h"
#include "auxiliary.h"
#include "ports.h"
#include "file_input_parsing.h"
#include "launch_manager.h"
#include "command_line_parsing.h"
#include "lexer.h"
#include "commonvars.h"
#include "dvda-author.h"
/*  Global  options */

globalData globals;
static char *LPLEXTEMPDIR;
char* TEMPDIR;

 command_t* lexer_analysis(command_t* command, lexer_t* lexer, const char* config_file, bool config_type)
 {
    int i;
    lexer->nlines=MAX_LEXER_LINES;

    lexer->commandline=(char** ) calloc(MAX_LEXER_LINES, sizeof(char *));
    if (lexer->commandline == NULL) perror("\n"ERR "lexer\n");
    for (i=0; i < MAX_LEXER_LINES; i++)
    {
        lexer->commandline[i]=(char* ) calloc(2*MAX_OPTION_LENGTH, sizeof(char));
        if (lexer->commandline[i] == NULL) perror("\n"ERR "lexer\n");
    }

    if (config_type == CONFIGURATION_FILE) check_settings_file();

    errno=0;

    config_lexer(config_file, lexer);

    if (command == NULL)
    {
        EXIT_ON_RUNTIME_ERROR_VERBOSE(ERR "Could not allocate command-line structure")
    }

    command=command_line_parsing(lexer->nlines, lexer->commandline, command);

    for (i=0; i < MAX_LEXER_LINES; i++)
        FREE(lexer->commandline[i])

        FREE(lexer->commandline)

        return command;
 }

static inline void allocate_paths(char** s, const char* dir, unsigned long length)
{
    if (s != NULL) free(*s);
    *s = calloc(length + 10, sizeof(char));
    sprintf(*s, "%s"SEPARATOR"%s", globals.settings.tempdir, dir);
}

void normalize_temporary_paths(pic* img)
{
   if (img == NULL) return;

   int menu;

   free(img->backgroundpic);
   img->backgroundpic=calloc(img->nmenus+1,sizeof(char*));
   free(img->backgroundmpg);
   img->backgroundmpg=calloc(img->nmenus+1,sizeof(char*));
   free(img->imagepic);
   img->imagepic=calloc(img->nmenus+1,sizeof(char*));
   free(img->highlightpic);
   img->highlightpic=calloc(img->nmenus+1,sizeof(char*));
   free(img->selectpic);
   img->selectpic=calloc(img->nmenus+1, sizeof(char*));

    // useless to realloc for just one menu !

   int s = strlen(globals.settings.tempdir);

   for (menu=0;  menu < img->nmenus; menu++)
    {
        free(img->backgroundmpg[menu]);
        img->backgroundmpg[menu]=(char*)calloc(26+s, sizeof(char));
        sprintf(img->backgroundmpg[menu], "%s"SEPARATOR"%s%d%s", globals.settings.tempdir, "background", menu, ".mpg");

        free(img->backgroundpic[menu]);
        img->backgroundpic[menu]=calloc(s+13, sizeof(char));
        sprintf(img->backgroundpic[menu], "%s"SEPARATOR"%s%d%s", globals.settings.tempdir, "bgpic", menu, ".jpg");

        img->imagepic[menu]=calloc(s+13, sizeof(char));
        sprintf(img->imagepic[menu], "%s"SEPARATOR"%s%d%s", globals.settings.tempdir, "impic", menu, ".png");

        free(img->highlightpic[menu]);
        img->highlightpic[menu]=calloc(s+13,sizeof(char));
        sprintf(img->highlightpic[menu], "%s"SEPARATOR"%s%d%s", globals.settings.tempdir, "hlpic", menu, ".png");

        img->selectpic[menu]=calloc(s+13,sizeof(char));
        sprintf(img->selectpic[menu], "%s"SEPARATOR"%s%d%s", globals.settings.tempdir, "slpic", menu, ".png");
    }

   img->imagepic[img->nmenus]=NULL;
   img->highlightpic[img->nmenus]=NULL;
   img->selectpic[img->nmenus]=NULL;
   img->backgroundmpg[img->nmenus]=NULL;
   globals.imagepicsize = img->nmenus + 1;
   globals.highlightpicsize = img->nmenus + 1;
   globals.backgroundmpgsize = img->nmenus + 1;
   globals.selectpicsize = img->nmenus + 1;
}


int main(int argc,  char* const argv[])
{

    int i=0;
    errno=0;

    if (errno) perror(ERR "Initial allocation\n");

    lexer_t   lexer_init;
    lexer_t   *lexer=&lexer_init;

    /****
     *  create a static command_t structure
     *  in command_line_parsing.c from default command line created by lexer.
     ****/

#ifndef __MINGW32__
    struct rusage nothing, start;
    compute_t timer= {&nothing, &start};
    starter(&timer);
#endif

    /*  DEFAULT   SETTINGS  */

    // Locale, time and log management

    setlocale(LC_ALL, LOCALE);

    char* currentdir = fn_get_current_dir_name();

    if (currentdir == NULL)
    {
      fprintf(stderr, "%s\n", ERR "Répertoire courant non alloué");
      clean_exit(-1);
    }

    int currentdirlength=strlen(currentdir);
    char TEMPDIRROOT[currentdirlength + 14];
    TEMPDIR = calloc(currentdirlength + 20, sizeof(char));
    LPLEXTEMPDIR = calloc(currentdirlength + 26, sizeof(char));
    char *EXECDIR = calloc(MAX(currentdirlength, 20) + 4 + 25, sizeof(char));  // /usr/local/bin or /usr/bin under *NIX, "currentdir" directory/bin otherwise (win32...)
    // 4 for "/bin and be liberal and allow 25 more characters for the executable name.

#ifdef BINDIR
    memcpy(EXECDIR, BINDIR, strlen(BINDIR));
#else
    sprintf(EXECDIR, "%s/bin", currentdir);
#endif

    char *DATADIR=strdup(currentdir);
    char ***SNDT=calloc(1, sizeof(char**));
    SNDT[0]=calloc(1, sizeof(char*));
    SNDT[0][0]=strdup(DEFAULT_SOUNDTRACK);
//    IMPIC[0]=calloc(currentdirlength+30, sizeof(char));
//    HLPIC[0]=calloc(currentdirlength+30, sizeof(char));
//    SLPIC[0]=calloc(currentdirlength+30, sizeof(char));


    sprintf(TEMPDIRROOT, "%s%s%s", currentdir,(currentdir[0] == 0)? "" : SEPARATOR , TEMPDIR_SUBFOLDER_PREFIX DVDA_AUTHOR_BASENAME);

    sprintf(TEMPDIR, "%s"SEPARATOR"%s", TEMPDIRROOT, "temp");
    sprintf(LPLEXTEMPDIR, "%s"SEPARATOR"%s", TEMPDIRROOT, "temp.lplex");


    // Global settings are hard-code set by default as follows:
    errno=0;

    /* requests C99 */

    defaults def = {
        .settingsfile = strdup(SETTINGSFILE),
        .logfile = NULL,   // logfile path should be supplied on command line
        .indir = NULL,
        .outdir = NULL,
        .outfile = NULL,
        .lplexoutdir = NULL,
    #ifdef __WIN32__
         .workdir = strdup(DEFAULT_WORKDIR),// working directory: under Windows, c:\ if not defined at compile time, otherwise 'currentdir' environment variable
    #else
         .workdir = strdup(currentdir),
    #endif
        .tempdir = NULL,
        .lplextempdir = NULL,
        .linkdir = NULL,
        .bindir =  EXECDIR, //bindir,
        .datadir = DATADIR,
        .fixwav_database = strdup(STANDARD_FIXWAV_DATABASE_PATH),
        .dvdisopath = NULL,
        .stillpicdir = NULL

    };

    uint32_t tab0[9] = {0};
    tab0[0] = 1;
    uint32_t tab1[9] = {0};
    tab1[0] = 1;
    uint32_t tab2[9] = {0};
    tab2[0] = 1;
    char**   tab3    = calloc(1, sizeof(char*));
    tab3[0] = strdup(DEFAULT_BACKGROUNDPIC);

    globalData globals_init=
    {

        /*top menu*/        NO_MENU,  // no top menu
        /*nooutput*/        false,
        /*runmkisofs*/      false,  // do not run mkisofs
        /*autoplay*/        false,  // no autoplay
        /*text table*/      false,  // no text table
        /*silence*/         false,
                            true,   // enabling lexer
        /*logfile*/	        false,  // no log
        /*loghtml*/         false,  // text log
        /* logdecode */     false,  // no forensic analysis of system files
        /*videozone*/       true,   // generates video zone
        /*videolinking*/    false,  // no video link
        /*decode */         false,  // decode MLP to WAV
        /*pipe*/   false,  // slow down extraction for buffering tp extrernal tools
        /*play*/            false,  // pipe to ffplay or other player
        /*playlist*/        false,  // no playlist
        /*cga*/             false,  // no explicit channel group assignement
        /*end_pause*/       false,  // no end pause
/*strict audio extraction */false,  // do not systematically check headers
        /*devel verbosity*/ false,  // no enhanced verbosity
        /*very verbose*/    false,  // not very verbose
        /*debugging*/       false,  // no debugging-level verbosity
        /*padding*/         false,  // gapless join for like audio tracks, no padding
        /*padding_continuous*/   false,  // no continuous padding
        /*lossy_rounding*/  false,  // No audio loss
        /*rungrowisofs*/    false,  // Do not burn with growisofs
#ifndef WITHOUT_sox
        /*sox_enable*/      false,  // no use of SoX
#endif

        /*fixwav_enable*/               true,  // use of fixwav
        /*fixwav_virtual_enable*/       true,  // use of fixwav (virtual headers)
        /* automatic behaviour */       true,
        /* do not prepend a header */   false,
        /* do not correct in place */   false,
        /* be cautious on overwrites*/  false,
        /* not interactive */           false,
        /* no padding */    false,
        /* prune */         false,
        /* force */         false,
        /* textablesize */  false,
        /* topmenusize */   false,
        .grouptextsize = tab1,
        .tracktextsize = tab2,
        .backgroundmpgsize = 0,
        .backgroundpicsize = 1,
        .soundtracksize =    tab0,
        /* topmenu_slidesize */ 0,
        /* highlightpicsize */  0,
        /* selectpicsize */     0,
        /* imagepicsize */      0,
        /* backgroundcolorssize */ 0,
        /* fixwav output suffix*/
        strdup(STANDARD_FIXWAV_SUFFIX),
        /*fixwav_parameters*/ NULL,
        /*xml filepath*/    NULL,
        /*spumux xml*/      NULL,
        /*cdrecord dev*/    NULL,
        /* aob path for decoding */ NULL,
       /*player*/            strdup(PLAYER),  // pipe to ffplay
        /*player-path*/            strdup(PLAYER_PATH),  // full path to player
        /*journal (log)*/   NULL, //(FILE*)
        /*access rights*/   DEFAULT_ACCESS_RIGHTS,
        /* it is necessary to use strdup as these settings may be overridden dynamically */
// Paths:

        def
    };

    pic     img0=
    {
        1, // always refresh menu pics by default (change of syntax, 10.06). Can be overridden with --colors norefresh
        0, // no loop
        0,  // list menus, not hierarchical
        0, // no active menus
        NULL,
        NULL,
        NULL,
        tab3, // black screen for jpg video mpg authoring
        strdup(DEFAULT_BLANKSCREEN), // black screen for png authoring
        NULL, //backgroundmpg
        NULL, //backgroundcolors
        strdup(DEFAULT_ACTIVEHEADER),
        NULL, //topmenu
        NULL,
        NULL, //stillvob
        NULL, //tsvob
        SNDT, //soundtrack  silence.wav
        strdup(DEFAULT_AUDIOFORMAT),
        strdup(DEFAULT_ALBUMCOLOR), //top menu pic album color
        strdup(DEFAULT_GROUPCOLOR), //top menu pic group color
        strdup(DEFAULT_ARROWCOLOR), //top menu pic arrow color
        strdup(DEFAULT_TEXTCOLOR_PIC), //top menu pic textcolor

        strdup(DEFAULT_BGCOLOR_PIC), //topmenu pic background color no strdup here (same-string optarg)
        strdup(DEFAULT_HCOLOR_PIC), //topmenu pic highlight color  no strdup here (same-string optarg)
        strdup(DEFAULT_SELCOLOR_PIC), //topmenu pic select action color no strdup here (same-string optarg)

        strdup(DEFAULT_ACTIVETEXTCOLOR_PALETTE), //top menu pic textcolor
        strdup(DEFAULT_ACTIVEBGCOLOR_PALETTE), //topmenu pic background color no strdup here (same-string optarg)
        strdup(DEFAULT_ACTIVEHCOLOR_PALETTE), //topmenu pic highlight color  no strdup here (same-string optarg)
        strdup(DEFAULT_ACTIVESELCOLOR_PALETTE), //topmenu pic select action color no strdup here (same-string optarg)

        strdup(DEFAULT_TEXTCOLOR_PALETTE), //system palette textcolor
        strdup(DEFAULT_BGCOLOR_PALETTE), //palette background color        no strdup here (same-string optarg)
        strdup(DEFAULT_HCOLOR_PALETTE), //palette highlight color    no strdup here (same-string optarg)
        strdup(DEFAULT_SELCOLOR_PALETTE), //palette select action color red   no strdup here (same-string optarg)

        strdup(DEFAULT_TEXTFONT), //textfont Courier
        NULL,
        strdup(PAL_FRAME_RATE),   // PAL frame rate by default
        strdup(PAL_NORM),
        "2", // 4:3 ratio
        strdup(DEFAULT_ASPECT_RATIO), // 4:3 ratio
        0, //  pointsize 30
        DEFAULT_FONTWIDTH, //  fontwidth 6
        UNDERLINE,
        0,  //h    // O length is usually OK
        0,  // min
        0,   // sec
        NOPICS,  // neither still pics nor video top menu
        0,   // no topmenu screen
        DEFAULT_MENU_NCOLUMNS,
        0,  // pic count
        NULL, // npics table
        NULL,
        NULL, // no still pic vob size
        NULL, // no menu vob size
        NULL // still pic options

    };

    globals=globals_init;
    globals.settings.tempdir=strdup(TEMPDIR);
    globals.settings.lplextempdir=LPLEXTEMPDIR;
    globals.settings.stillpicdir=strdup(globals.settings.tempdir);
    unsigned long s = strlen(globals.settings.tempdir);

    allocate_paths(&globals.settings.indir, "audio", s);
    allocate_paths(&globals.settings.outdir, "output", s);
    allocate_paths(&globals.settings.lplexoutdir, "output", s);

    // Null arg is no longer supported, yet...

    if (argc == 1)
    {
        foutput("\n%s", "dvda-author syntax:\n------------------\n");
        help();
        return(errno);
    }

    // Now we are sure argc > 0

    // If a default setting textfile exists, it overrides the above hard-coded defaults
    // Path to defaults settings file must be "Path/to/DVD-A author folder/dvda-author.conf"
    // or otherwise defined by symbolic variable SETTINGSFILE at compile time.
    // Default settings, either hard-code set or by defaults textfile, can be overridden by command-line.

    /* Lexer extracts default command line from dvda-author.conf and returns corresponding argc, argv
       yet this is useless if command line is just "--version" or "--help" or equivalents */


    bool project_flag=0;
    char* project_filepath=NULL;
    command_t command0, *command=NULL;
    command0.img=&img0;
    command=&command0;

    for (i=1; i < argc ; i++)
        if (strcmp(argv[i],"--disable-lexer") == 0 || strcmp(argv[i] , "-W") ==0 || strcmp(argv[i], "--version") ==0 || strcmp(argv[i], "--help") ==0 || strcmp(argv[i], "-v") ==0 || strcmp(argv[i], "-h") == 0)
        {
            globals.enable_lexer=0;
                goto launch;
        }
        else
        if (strcmp(argv[i], "--project") == 0)
        {
           project_flag=1;
           if (i+1 < argc && argv[i+1][0] != '-')
               project_filepath=strdup(argv[i+1]);
           else
              project_filepath=strdup(DEFAULT_DVDA_AUTHOR_PROJECT_FILENAME);

           path_t *pstruct=parse_filepath(project_filepath);
           if (pstruct && pstruct->isfile)
           {
             //if (globals.debugging)
             foutput(INF "Parsing project file %s\n", project_filepath);
           }
           else
           {
            foutput(ERR "Failed to parse project file %s\n       Exiting...\n", project_filepath);
            clean_exit(EXIT_FAILURE);
           }
           free(pstruct);
        }

    lexer_analysis(command, lexer, SETTINGSFILE, CONFIGURATION_FILE);

        /* launch core processes after parsing user command-line, possibly overriding defaut values */


launch:

        if (project_flag)

          launch_manager(lexer_analysis(command, lexer, project_filepath, PROJECT_FILE));
        else

          launch_manager(command_line_parsing(argc, argv, command));

    // allocated in command_line_parsing()

    /* Compute execution time and exit */

    COMPUTE_EXECTIME

    FREE(currentdir)

    fflush(NULL);
    if ((globals.loghtml) && (globals.logfile)) htmlize(globals.settings.logfile);
    if ((globals.logfile) && (globals.journal)) fclose(globals.journal);

    if (globals.end_pause) pause_dos_type();

    if (errno && globals.veryverbose)
    {
        perror(WAR "Detected runtime errors");
    }

    return(errno);
}

